### PR TITLE
add error in the first camera to the cost function

### PR DESCRIPTION
**Pull request by lpanaf**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by ivashmak** [comment]
> @ivashmak Could you take a look too?

I think it's a good idea to add error for the first camera too. Code-wise, everything seems right.

File: modules/calib/src/multiview_calibration.cpp [hunk 1]
```diff
@@ -382,50 +382,49 @@ static void optimizeLM (std::vector<double> &param, const RobustFunction &robust
...
-                    {
-                        Matx13d de3dr3, de3dt3, de3dr2, de3dt2, de3dr1, de3dt1;
-                        for (int j = 0; j < 3; j++)
-                            de3dr3(j) = Je.at<double>(p, j);
 
-                        for (int j = 0; j < 3; j++)
-                            de3dt3(j) = Je.at<double>(p, 3 + j);
+                    Mat wd;
+                    Mat::diag(weights).convertTo(wd, CV_64F);
+                    if (k > 0) { // if not the first frame
```
Comment (asmorkalov) [inline modules/calib/src/multiview_calibration.cpp:390]: if not the first frame -> if not the first camera

**Comment by asmorkalov** [comment]
Please remove trailing whitespaces:
```
modules/calib/src/multiview_calibration.cpp:390: trailing whitespace.
+                    if (k > 0) { // if not the first frame 
modules/calib/src/multiview_calibration.cpp:422: trailing whitespace.
+                        
modules/calib/src/multiview_calibration.cpp:425: trailing whitespace.
+                        JtJ(Rect(eofs, (k - 1) * 6, 6, 6)) = (J_0ToK.t() * wd * Je); 
```

File: modules/calib/src/multiview_calibration.cpp [hunk 2]
```diff
@@ -328,12 +328,12 @@ static void optimizeLM (std::vector<double> &param, const RobustFunction &robust
         cnt_valid_frame = 0;
         for (int i = 0; i < NUM_FRAMES; i++ ) {
             if (!valid_frames[i]) continue;
-            for (int k = 1; k < NUM_CAMERAS; k++ ) { // skip first camera as there is nothing to optimize
+            for (int k = 0; k < NUM_CAMERAS; k++ ) { // for the first camera, take caution that only the frame camera is to be optimized
```
Comment (asmorkalov) [inline modules/calib/src/multiview_calibration.cpp:331]: The comment is not clear. I propose to state: Pose for camera #0 is not optimized (coordinates center), but it's re-projection error is taken into account.

**Review by asmorkalov — COMMENTED** [review]
Looks good in general for me.

**Comment by asmorkalov** [comment]
@lpanaf Could you attach your experimental numbers with synthetic data and how they are got. I'll make several experiments and return back with update too.

**Comment by asmorkalov** [comment]
@ivashmak What do you think about the patch?

**Comment by ivashmak** [comment]
> @ivashmak What do you think about the patch?

I think this idea should work. In the previous version, the first camera was not included in the optimization, because in perfect scenario extrinsics are fixed to identity and zero and shouldn't be optimized. However, since the optimization is additionally done over rigid rotation of the pattern grid for each frame (ie, om and T) then there is a small increase in the reprojection error for the first camera at each iteration, because om and T are now optimized for all cameras except the first one. Therefore, adding a reprojection loss for the first camera will help the optimization to fit om and T parameters equally for all cameras.

I have looked at the code, and it seems reasonable for me to work, but I have not tested it myself. If my help is needed, I could do that.

**Comment by asmorkalov** [comment]
My experiments with 3-4 cameras shows small reduction of mean RMS over all visible frames and per-camera RMS reduction. So, we can treat the patch as useful. @lpanaf Could you fix the remarks in code and I'll merge the patch.

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**